### PR TITLE
selection and cursor pos

### PR DESCRIPTION
**Pull request by Sublustris**
Need some help in  rich_input.lua   -  line 186

### CODE REVIEW

**Comment by Sublustris** [comment]
That's all the features we want to add this time,  please let us know when you would be ready with code review. If something needs fixes, let us know too pls.

File: druid/custom/rich_input/rich_input.lua [hunk 1]
```diff
@@ -71,15 +140,26 @@ function RichInput.init(self, template, nodes)
...
 
 	self.input.on_input_text:subscribe(update_text)
 	self.input.on_input_select:subscribe(on_select)
 	self.input.on_input_unselect:subscribe(on_unselect)
-	on_unselect(self)
-	update_text(self, "")
+	
+	self.input.button.on_click:subscribe(on_button_click, self)
+	self.input.button.on_double_click:subscribe(on_button_double_click, self)
+	self.input.style.NO_CONSUME_INPUT_WHILE_SELECTED = true
```
Comment (Insality) [inline druid/custom/rich_input/rich_input.lua:154]: I think the rich input should not change these style settings (and for skip input keys same)

File: druid/custom/rich_input/rich_input.lua [hunk 2]
```diff
@@ -30,37 +31,105 @@ local SCHEME = {
...
+local function clear_text(self, replace_with_symbol)
+	local spacer = replace_with_symbol or ""
+	self.input:set_text(spacer)
+	update_text(self)
+
+	if replace_with_symbol then
+		gui.set_position(self.cursor, vmath.vector3(self.input.total_width/2, 0, 0))
+		self.input.cursor_letter_index = 1
+	else
+		gui.set_position(self.cursor, vmath.vector3(0, 0, 0))
```
Comment (Insality) [inline druid/custom/rich_input/rich_input.lua:92]: You can create a `local TEMP_VECTOR = vector3(0)` in module scope, change their fields instead creating vector each time. It allows to save a bit of memory

File: druid/custom/rich_input/rich_input.lua [hunk 3]
```diff
@@ -92,4 +172,55 @@ function RichInput.set_placeholder(self, placeholder_text)
 end
 
 
+function RichInput.on_input(self, action_id, action)
+	self.action_pos_x = action.screen_x
```
Comment (Insality) [inline druid/custom/rich_input/rich_input.lua:176]: The values `action_pos_x` and `touch_pos_x` should not be stored in rich input component. You can use it as a local variables inside only on_input functions

File: druid/custom/rich_input/rich_input.lua [hunk 4]
```diff
@@ -92,4 +172,55 @@ function RichInput.set_placeholder(self, placeholder_text)
...
+			if self.input.cursor_letter_index > 0 then 
+				local text = self.input:get_text()
+				local new_text = utf8.sub(text, 1, self.input.cursor_letter_index-1) .. utf8.sub(text, self.input.cursor_letter_index +1 ) 
+				self.input.cursor_letter_index = self.input.cursor_letter_index -1
+				self.input:set_text(new_text)
+			end
+		end
+		
+		if action_id == const.ACTION_TEXT then
+			--self.input.cursor_letter_index = self.input.cursor_letter_index +1
```
Comment (Insality) [inline druid/custom/rich_input/rich_input.lua:201]: Empty if + comments

File: druid/extended/input.lua [hunk 5]
```diff
@@ -219,10 +228,12 @@ function Input.on_input(self, action_id, action)
 
 		if input_text or #self.marked_value > 0 then
 			self:set_text(input_text)
-			return true
+			--return true
```
Comment (Insality) [inline druid/extended/input.lua:231]: Looks like a mistake

**Review by Insality — CHANGES_REQUESTED** [review]
Thanks for the changes!

Let's first fix all UX things, like more accurate selection and input handling (currently the rich input consumes all the input)

**Pull request closed without merging**